// apify-client/build/apify_error.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.NOT_FOUND_STATUS_CODE = exports.REQUEST_FAILED_ERROR_MESSAGE = exports.REQUEST_FAILED_ERROR_TYPE = exports.INVALID_PARAMETER_ERROR_TYPE = exports.APIFY_ERROR_NAME = undefined;

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const APIFY_ERROR_NAME = exports.APIFY_ERROR_NAME = 'ApifyClientError';

const INVALID_PARAMETER_ERROR_TYPE = exports.INVALID_PARAMETER_ERROR_TYPE = 'invalid-parameter';
const REQUEST_FAILED_ERROR_TYPE = exports.REQUEST_FAILED_ERROR_TYPE = 'request-failed';
const REQUEST_FAILED_ERROR_MESSAGE = exports.REQUEST_FAILED_ERROR_MESSAGE = 'Server request failed.';
const NOT_FOUND_STATUS_CODE = exports.NOT_FOUND_STATUS_CODE = 404;

class ApifyClientError extends Error {
    constructor(type, message, details) {
        super(message);
        this.name = APIFY_ERROR_NAME;
        this.type = type;
        this.details = details;

        Error.captureStackTrace(this, ApifyClientError);
    }

    toString() {
        let details = '';

        if (this.details) {
            const items = [];
            _underscore2.default.mapObject(this.details, (val, key) => {
                items.push(`${key}=${JSON.stringify(val)}`);
            });
            details = `(${items.join(', ')})`;
        }

        return `[${this.name}] ${APIFY_ERROR_NAME}: ${this.message} ${details}`;
    }
}
exports.default = ApifyClientError;
