// apify-client/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

var _utils = require('./utils');

var _acts = require('./acts');

var _acts2 = _interopRequireDefault(_acts);

var _tasks = require('./tasks');

var _tasks2 = _interopRequireDefault(_tasks);

var _key_value_stores = require('./key_value_stores');

var _key_value_stores2 = _interopRequireDefault(_key_value_stores);

var _datasets = require('./datasets');

var _datasets2 = _interopRequireDefault(_datasets);

var _logs = require('./logs');

var _logs2 = _interopRequireDefault(_logs);

var _users = require('./users');

var _users2 = _interopRequireDefault(_users);

var _webhooks = require('./webhooks');

var _webhooks2 = _interopRequireDefault(_webhooks);

var _webhook_dispatches = require('./webhook_dispatches');

var _webhook_dispatches2 = _interopRequireDefault(_webhook_dispatches);

var _request_queues = require('./request_queues');

var _request_queues2 = _interopRequireDefault(_request_queues);

var _apify_error = require('./apify_error');

var _apify_error2 = _interopRequireDefault(_apify_error);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @ignore */
const getDefaultOptions = () => ({
    baseUrl: 'https://api.apify.com'
});

/**
 * IMPORTANT:
 *
 * This file MUST contain only one export which is default export of Apify Client.
 * Otherwise it would not get exported under require('apify-client') but ugly
 * require('apify-client').default instead.
 *
 * See: https://github.com/59naga/babel-plugin-add-module-exports
 */

/**
 * Each property is a plain object of methods.
 *
 * Method must have 2 parameters "options" and "requestPromise" where:
 * - requestPromise is utils.requestPromise with Promise parameter set based on
 *   user's choice of Promises dependency.
 * - options
 *
 * Method must return promise.
 * @ignore
 */
const methodGroups = {
    acts: _acts2.default,
    tasks: _tasks2.default,
    keyValueStores: _key_value_stores2.default,
    datasets: _datasets2.default,
    requestQueues: _request_queues2.default,
    logs: _logs2.default,
    users: _users2.default,
    webhooks: _webhooks2.default,
    webhookDispatches: _webhook_dispatches2.default
};

/**
 * @type package
 * @class ApifyClient
 * @param {Object} [options] - Global options for ApifyClient. You can globally configure here any method option from any namespace. For example
 *                             if you are working with just one actor then you can preset it's actId here instead of passing it to each
 *                             actor's method.
 * @param {String} [options.userId] - Your user ID at apify.com
 * @param {String} [options.token] - Your API token at apify.com
 * @param {Number} [options.expBackOffMillis=500] - Wait time in milliseconds before repeating request to Apify API in a case of server
                                                    or rate limit error
 * @param {Number} [options.expBackOffMaxRepeats=8] - Maximum number of repeats in a case of error
 * @param {Array<Number>} [options.retryOnStatusCodes=[429]] - An array of status codes on which request gets retried. By default requests are retried
 *                                                             only in a case of limit error (status code 429).
 * @description Basic usage of ApifyClient:
 * ```javascript
 * const ApifyClient = require('apify-client');
 *
 * const apifyClient = new ApifyClient({
 *   userId: 'jklnDMNKLekk',
 *   token: 'SNjkeiuoeD443lpod68dk',
 * });
 * ```
 *
 * All API calls done through this client are made with exponential backoff.
 * What this means, is that if the API call fails, this client will attempt the call again with a small delay.
 * If it fails again, it will do another attempt after twice as long and so on, until one attempt succeeds
 * or 8th attempt fails.
 */
const ApifyClient = function (options = {}) {
    // This allows to initiate ApifyClient both ways - with and without "new".
    if (!this || this.constructor !== ApifyClient) return new ApifyClient(options);

    // This is used only internally for unit testing of ApifyClient.
    const undecoratedMethodGroups = options._overrideMethodGroups || methodGroups;
    delete options._overrideMethodGroups;

    const instanceOpts = Object.assign({}, getDefaultOptions(), options);

    /**
     * This decorator does:
     * - extends "options" parameter with values from default options and from ApifyClient instance options
     * - adds options.baseUrl
     * - passes preconfigured utils.requestPromise
     * - allows to use method with both callbacks and promises
     * @ignore
     */
    const methodDecorator = method => {
        return (callOpts, callback) => {
            const mergedOpts = Object.assign({}, instanceOpts, callOpts);

            // Check that all required parameters are set.
            if (!instanceOpts.baseUrl) throw new _apify_error2.default(_apify_error.INVALID_PARAMETER_ERROR_TYPE, 'The "options.baseUrl" parameter is required');

            // Remove traling forward slash from baseUrl.
            if (mergedOpts.baseUrl.substr(-1) === '/') mergedOpts.baseUrl = mergedOpts.baseUrl.slice(0, -1);

            const preconfiguredRequestPromise = requestPromiseOptions => {
                return (0, _utils.requestPromise)(Object.assign({}, _underscore2.default.pick(mergedOpts, _utils.REQUEST_PROMISE_OPTIONS), requestPromiseOptions), this.stats);
            };

            const promise = method(preconfiguredRequestPromise, mergedOpts);
            if (!callback) return promise;

            promise.then(result => callback(null, result), error => callback(error));
        };
    };

    // Decorate methods and bind them to this object.
    _underscore2.default.forEach(undecoratedMethodGroups, (methodGroup, name) => {
        this[name] = _underscore2.default.mapObject(methodGroup, methodDecorator);
    });
    /**
     * Overrides options of ApifyClient instance.
     * @memberof ApifyClient
     * @function setOptions
     * @instance
     * @param {Object} options - See {@link ApifyClient} options object
     */
    this.setOptions = newOptions => {
        _underscore2.default.forEach(newOptions, (val, key) => {
            instanceOpts[key] = val;
        });
    };

    /**
     * Returns options of ApifyClient instance.
     * @memberof ApifyClient
     * @function getOptions
     * @instance
     * @return {Object} See {@link ApifyClient} constructor options
     */
    this.getOptions = () => {
        return _underscore2.default.clone(instanceOpts);
    };

    /**
     * This helper function is used in unit tests.
     * @ignore
     */
    this.getDefaultOptions = getDefaultOptions;

    /**
     * An object that contains various statistics about the API operations.
     * @memberof ApifyClient
     * @instance
     */
    this.stats = {
        // Number of Apify client function calls
        calls: 0,

        // Number of Apify API requests
        requests: 0,

        // Number of times the API returned 429 error. Spread based on number of retries.
        rateLimitErrors: new Array(Math.max(_request_queues.REQUEST_ENDPOINTS_EXP_BACKOFF_MAX_REPEATS, _utils.EXP_BACKOFF_MAX_REPEATS)).fill(0)

        // TODO: We can add internalServerErrors and other stuff here...
    };
};

exports.default = ApifyClient;
module.exports = exports.default;
