// apify-client/build/tasks.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.BASE_PATH = undefined;

var _underscore = require('underscore');

var _underscore2 = _interopRequireDefault(_underscore);

var _log = require('apify-shared/log');

var _log2 = _interopRequireDefault(_log);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Tasks
 * @memberOf ApifyClient
 * @description
 * ### Basic usage
 * Every method can be used as either promise or with callback. If your Node version supports await/async then you can await promise result.
 * ```javascript
 * const ApifyClient = require('apify-client');
 *
 * const apifyClient = new ApifyClient({
 *  userId: 'jklnDMNKLekk',
 *  token: 'SNjkeiuoeD443lpod68dk',
 * });
 *
 * // Awaited promise
 * try {
 *      const tasksList = await apifyClient.tasks.listTasks({});
 *      // Do something with the tasksList ...
 * } catch (err) {
 *      // Do something with error ...
 * }
 *
 * // Promise
 * apifyClient.tasks.listTasks({})
 * .then((tasksList) => {
 *      // Do something tasksList ...
 * })
 * .catch((err) => {
 *      // Do something with error ...
 * });
 *
 * // Callback
 * apifyClient.tasks.listTasks({}, (err, tasksList) => {
 *      // Do something with error or tasksList ...
 * });
 * ```
 * @namespace tasks
 */

const BASE_PATH = exports.BASE_PATH = '/v2/actor-tasks';

const replaceSlashWithTilde = str => str.replace('/', '~');

exports.default = {
    /**
     * Gets list of your tasks.
     * @description By default, the objects are sorted by the createdAt field in ascending order,
     * therefore you can use pagination to incrementally fetch all tasks while new ones are still being created.
     * To sort them in descending order, use desc: `true` parameter.
     * The endpoint supports pagination using limit and offset parameters and it will not return more than 1000 array elements.
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {Number} [options.offset=0] - Number of array elements that should be skipped at the start.
     * @param {Number} [options.limit=1000] - Maximum number of array elements to return.
     * @param {Boolean} [options.desc] - If `true` then the objects are sorted by the createdAt field in descending order.
     * @param callback
     * @returns {PaginationList}
     */
    listTasks: (requestPromise, options) => {
        const { baseUrl, token, offset, limit, desc } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(limit, 'limit', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(offset, 'offset', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(desc, 'desc', 'Maybe Boolean');

        const query = { token };

        if (limit) query.limit = limit;
        if (offset) query.offset = offset;
        if (desc) query.desc = 1;

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}`,
            json: true,
            method: 'GET',
            qs: query
        }).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Creates a new task.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {Object} options.task Object containing configuration of the task
     * @param callback
     * @returns {Task}
     */
    createTask: (requestPromise, options) => {
        const { baseUrl, token, task } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(task, 'task', 'Object');

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}`,
            json: true,
            method: 'POST',
            qs: { token },
            body: task
        }).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Updates task.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {String} options.taskId - Unique task ID
     * @param {Object} options.task
     * @param callback
     * @returns {Task}
     */
    updateTask: (requestPromise, options) => {
        const { baseUrl, token, taskId, task } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(task, 'task', 'Object');

        if (taskId) (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');else (0, _utils.checkParamOrThrow)(task.id, 'task.id', 'String');

        const safeTaskId = replaceSlashWithTilde(!taskId && task.id ? task.id : taskId);

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}`,
            json: true,
            method: 'PUT',
            qs: { token },
            body: _underscore2.default.omit(task, 'id')
        }).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Deletes task.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {String} options.taskId - Unique task ID
     * @param callback
     */
    deleteTask: (requestPromise, options) => {
        const { baseUrl, token, taskId } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');

        const safeTaskId = replaceSlashWithTilde(taskId);

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}`,
            json: true,
            method: 'DELETE',
            qs: { token }
        }).then(_utils.parseDateFields);
    },

    /**
     * Gets task object.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param {String} options.token Optional
     * @param {String} options.taskId - Unique task ID
     * @param callback
     * @returns {Task}
     */
    getTask: (requestPromise, options) => {
        const { baseUrl, token, taskId } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        // Remove line bellow when we enable public tasks
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');

        const safeTaskId = replaceSlashWithTilde(taskId);

        const qs = {};
        // Line bellow will be used if we enable tasks to be public
        // if (token) qs.token = token;
        qs.token = token;

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}`,
            json: true,
            method: 'GET',
            qs
        }).then(_utils.pluckData).then(_utils.parseDateFields).catch(_utils.catchNotFoundOrThrow);
    },

    /**
     * Gets list of task runs.
     *
     * By default, the objects are sorted by the startedAt field in ascending order,
     * therefore you can use pagination to incrementally fetch all builds while new ones are still being created.
     * To sort them in descending order, use desc: `true` parameter.
     *
     * The endpoint supports pagination using limit and offset parameters and it will not return more than 1000 array elements.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {String} options.taskId - Unique task ID
     * @param {Number} [options.offset=0] - Number of array elements that should be skipped at the start.
     * @param {Number} [options.limit=1000] - Maximum number of array elements to return.
     * @param {Boolean} [options.desc] - If `true` then the objects are sorted by the createdAt field in descending order.
     * @param callback
     * @returns {PaginationList}
     */
    listRuns: (requestPromise, options) => {
        const { baseUrl, token, taskId, offset, limit, desc } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');
        (0, _utils.checkParamOrThrow)(limit, 'limit', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(offset, 'offset', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(desc, 'desc', 'Maybe Boolean');

        const safeTaskId = replaceSlashWithTilde(taskId);
        const query = { token };

        if (limit) query.limit = limit;
        if (offset) query.offset = offset;
        if (desc) query.desc = 1;

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}/runs`,
            json: true,
            method: 'GET',
            qs: query
        }).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Runs the given task.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param {String} options.taskId - Unique task ID
     * @param [options.token]
     * @param {Number} [options.waitForFinish] - Number of seconds to wait for task to finish. Maximum value is 120s.
                                                 If task doesn't finish in time then task run in RUNNING state is returned.
     * @param {Object} [options.input] - Actor task input object.
     * @param {Number} [options.timeout] - Timeout for the act run in seconds. Zero value means there is no timeout.
     * @param {Number} [options.memory] - Amount of memory allocated for the act run, in megabytes.
     * @param {String} [options.build] - Tag or number of the build to run (e.g. <code>latest</code> or <code>1.2.34</code>).
     * @param {Array}  [options.webhooks] - Specifies optional webhooks associated with the actor run,
     *                                      which can be used to receive a notification e.g. when the actor finished or failed,
     *                                      see {@link https://docs.apify.com/webhooks/ad-hoc-webhooks|ad hook webhooks documentation}
     *                                      for detailed description.
     * @param callback
     * @returns {ActRun}
     */
    runTask: (requestPromise, options) => {
        const { baseUrl, token, taskId, waitForFinish, body, contentType, timeout, memory, build, webhooks, input } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');
        (0, _utils.checkParamOrThrow)(waitForFinish, 'waitForFinish', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(timeout, 'timeout', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(memory, 'memory', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(build, 'build', 'Maybe String');
        (0, _utils.checkParamOrThrow)(webhooks, 'webhooks', 'Maybe Array');
        (0, _utils.checkParamOrThrow)(input, 'input', 'Maybe Object');

        const safeTaskId = replaceSlashWithTilde(taskId);
        const query = { token };

        if (waitForFinish) query.waitForFinish = waitForFinish;
        if (timeout) query.timeout = timeout;
        if (memory) query.memory = memory;
        if (build) query.build = build;
        if (webhooks) query.webhooks = (0, _utils.stringifyWebhooksToBase64)(webhooks);

        const opts = {
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}/runs`,
            method: 'POST',
            qs: query
        };

        // This is for backwards compatiblity.
        // TODO: Remove when releasing v1.0.
        if (body) {
            if (input) {
                throw new Error('You cannot use deprecated parameter "body" along with its replacement "input"!');
            }
            (0, _utils.checkParamOrThrow)(contentType, 'contentType', 'Maybe String');
            (0, _utils.checkParamOrThrow)(body, 'body', 'String');
            _log2.default.deprecated('Parameter "body" of client.tasks.runTask() method is depredated. Use "input" parameter instead.');
            opts.body = body;
            if (contentType) opts.headers = { 'Content-Type': contentType };
        }

        if (input) {
            opts.body = input;
            opts.json = true;
        }

        return requestPromise(opts).then(response => opts.json ? response : JSON.parse(response)).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Gets list of webhooks for given actor task.
     *
     * @memberof ApifyClient.acts
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {String} options.taskId - Unique task ID
     * @param {Number} [options.offset=0] - Number of array elements that should be skipped at the start.
     * @param {Number} [options.limit=1000] - Maximum number of array elements to return.
     * @param {Boolean} [options.desc] - If `true` then the objects are sorted by the createdAt field in descending order.
     * @param callback
     * @returns {PaginationList}
     */
    listWebhooks: (requestPromise, options) => {
        const { baseUrl, token, taskId, offset, limit, desc } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(limit, 'limit', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(offset, 'offset', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(desc, 'desc', 'Maybe Boolean');

        const safeTaskId = replaceSlashWithTilde(taskId);
        const query = { token };

        if (limit) query.limit = limit;
        if (offset) query.offset = offset;
        if (desc) query.desc = 1;

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}/webhooks`,
            json: true,
            method: 'GET',
            qs: query
        }).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Gets the actor task input.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {String} options.taskId - Unique task ID
     * @param callback
     * @returns {Object}
     */
    getInput: (requestPromise, options) => {
        const { baseUrl, token, taskId } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');

        const safeTaskId = replaceSlashWithTilde(taskId);

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}/input`,
            json: true,
            method: 'GET',
            qs: { token }
        });
    },

    /**
     * Updates the actor task input.
     *
     * @memberof ApifyClient.tasks
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {String} options.taskId - Unique task ID
     * @param {Object} options.input - Input object.
     * @param callback
     * @returns {Object}
     */
    updateInput: (requestPromise, options) => {
        const { baseUrl, token, taskId, input } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(taskId, 'taskId', 'String');
        (0, _utils.checkParamOrThrow)(input, 'input', 'Object');

        const safeTaskId = replaceSlashWithTilde(taskId);

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${safeTaskId}/input`,
            json: true,
            method: 'PUT',
            qs: { token },
            body: input
        });
    }
};
