// apify-client/build/users.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.BASE_PATH = undefined;

var _consts = require('apify-shared/consts');

var _utils = require('./utils');

/**
 * Users
 * @memberOf ApifyClient
 * @namespace users
 */

const BASE_PATH = exports.BASE_PATH = '/v2/users';

exports.default = {
    /**
     * Returns private and public information about user account.
     *
     * @memberof ApifyClient.users
     * @instance
     * @param {Object} options
     * @param {String} [options.token] - If set, the function returns a private and public information for the user account,
     *                                   otherwise it only returns public information.
     * @param {String} [options.userId = 'me'] - Desired user ID or username. By default it is `'me'`,
     *                                           which causes the function to return information about the current user.
     * @param {function} [callback] - Callback function
     * @returns {UserInfo}
     */
    getUser: (requestPromise, options) => {
        const { baseUrl, userId = _consts.ME_USER_NAME_PLACEHOLDER, token } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(userId, 'userId', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'Maybe String');

        const requestOpts = {
            url: `${baseUrl}${BASE_PATH}/${userId}`,
            json: true,
            method: 'GET'
        };

        if (token) requestOpts.qs = { token };

        return requestPromise(requestOpts).then(_utils.pluckData).then(_utils.parseDateFields);
    }
};
