// apify-client/build/webhook_dispatches.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.BASE_PATH = undefined;

var _utils = require('./utils');

/**
 * Webhook dispatches
 * @memberOf ApifyClient
 * @description
 * ### Basic usage
 * Every method can be used as either promise or with callback. If your Node version supports await/async then you can await promise result.
 * ```javascript
 * const ApifyClient = require('apify-client');
 *
 * const apifyClient = new ApifyClient({
 *  userId: 'jklnDMNKLekk',
 *  token: 'SNjkeiuoeD443lpod68dk',
 * });
 *
 * // Awaited promise
 * try {
 *      const dispatchesList = await apifyClient.webhookDispatches.listWebhookDispatches({});
 *      // Do something with the list ...
 * } catch (err) {
 *      // Do something with error ...
 * }
 *
 * // Promise
 * apifyClient.webhookDispatches.listWebhookDispatches({})
 * .then((webhooksList) => {
 *      // Do something with the list ...
 * })
 * .catch((err) => {
 *      // Do something with error ...
 * });
 *
 * // Callback
 * apifyClient.webhookDispatches.listWebhookDispatches({}, (err, webhooksList) => {
 *      // Do something with error or list ...
 * });
 * ```
 * @namespace webhookDispatches
 */
const BASE_PATH = exports.BASE_PATH = '/v2/webhook-dispatches';

exports.default = {
    /**
     * Gets list of webhook dispatches.
     * @description By default, the objects are sorted by the startedAt field in ascending order,
     * therefore you can use pagination to incrementally fetch all dispatches.
     * To sort them in descending order, use desc: `true` parameter.
     * The endpoint supports pagination using limit and offset parameters and it will not return more than 1000 array elements.
     * @memberof ApifyClient.webhookDispatches
     * @instance
     * @param {Object} options
     * @param options.token
     * @param {Number} [options.offset=0] - Number of array elements that should be skipped at the start.
     * @param {Number} [options.limit=1000] - Maximum number of array elements to return.
     * @param {Boolean} [options.desc] - If `true` then the objects are sorted by the startedAd field in descending order.
     * @param callback
     * @returns {PaginationList}
     */
    listDispatches: (requestPromise, options) => {
        const { baseUrl, token, offset, limit, desc } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');
        (0, _utils.checkParamOrThrow)(limit, 'limit', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(offset, 'offset', 'Maybe Number');
        (0, _utils.checkParamOrThrow)(desc, 'desc', 'Maybe Boolean');

        const query = { token };

        if (limit) query.limit = limit;
        if (offset) query.offset = offset;
        if (desc) query.desc = 1;

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}`,
            json: true,
            method: 'GET',
            qs: query
        }).then(_utils.pluckData).then(_utils.parseDateFields);
    },

    /**
     * Gets webhook dispatch.
     * @memberof ApifyClient.webhookDispatches
     * @instance
     * @param {Object} options
     * @param options.token
     * @param options.webhookDispatchId - Webhook dispatch ID
     * @param callback
     * @returns {WebhookDispatch}
     */
    getDispatch: (requestPromise, options) => {
        const { baseUrl, token, webhookDispatchId } = options;

        (0, _utils.checkParamOrThrow)(baseUrl, 'baseUrl', 'String');
        (0, _utils.checkParamOrThrow)(webhookDispatchId, 'webhookDispatchId', 'String');
        (0, _utils.checkParamOrThrow)(token, 'token', 'String');

        return requestPromise({
            url: `${baseUrl}${BASE_PATH}/${webhookDispatchId}`,
            json: true,
            method: 'GET',
            qs: { token }
        }).then(_utils.pluckData).then(_utils.parseDateFields).catch(_utils.catchNotFoundOrThrow);
    }
};
