// apify-shared/logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _events = _interopRequireDefault(require("events"));

var _log_consts = require("./log_consts");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This is an abstract class that should
 * be extended by custom logger classes.
 *
 * this._log() method must be implemented by them.
 */
class Logger extends _events.default {
  constructor(options) {
    super();
    this.options = options;
  }

  setOptions(options) {
    this.options = Object.assign({}, this.options, options);
  }

  getOptions() {
    return this.options;
  }

  _outputWithConsole(level, line) {
    switch (level) {
      case _log_consts.LEVELS.ERROR:
        console.error(line);
        break;

      case _log_consts.LEVELS.WARNING:
        console.warn(line);
        break;

      case _log_consts.LEVELS.DEBUG:
        console.debug(line);
        break;

      default:
        console.log(line);
    }
  }

  _log() {
    throw new Error('log() method must be implemented!');
  }

  log(...args) {
    const line = this._log(...args);

    this.emit('line', line);
  }

}

exports.default = Logger;
module.exports = exports.default;
